### PR TITLE
[feat] hard negative sampler

### PR DESCRIPTION
**Pull request by eric-gecheng**
Support hard negative samplers for match models.

### CODE REVIEW

File: tzrec/datasets/data_parser.py [hunk 1]
```diff
@@ -196,6 +196,10 @@ def parse(self, input_data: Dict[str, pa.Array]) -> Dict[str, torch.Tensor]:
                     f"sample weight column [{weight_name}] should be float dtype."
                 )
 
+        if "hard_neg_indices" in input_data.keys():
```
Comment (tiankongdeguiji) [inline tzrec/datasets/data_parser.py:204]: Making "hard_neg_indices" a constant value is better

File: tzrec/datasets/data_parser.py [hunk 2]
```diff
@@ -379,6 +383,13 @@ def to_batch(
         for weight in self._sample_weights:
             sample_weights[weight] = input_data[weight]
 
+        hard_neg_indices = torch.tensor([])
+        if (
```
Comment (tiankongdeguiji) [inline tzrec/datasets/data_parser.py]: input_data is a dict,  if "hard_neg_indices" in input_data.keys():

File: tzrec/main.py [hunk 3]
```diff
@@ -214,11 +214,24 @@ def _get_dataloader(
     return dataloader
 
 
+def _get_sampler_type(data_config: DataConfig) -> Optional[str]:
+    try:
```
Comment (tiankongdeguiji) [inline tzrec/main.py:218]: why try catch?

File: tzrec/main.py [hunk 4]
```diff
@@ -963,11 +986,13 @@ def export(
         data_config, features, pipeline_config.train_input_path, mode=Mode.PREDICT
     )
 
+    sampler_type = _get_sampler_type(data_config)
```
Comment (tiankongdeguiji) [inline tzrec/main.py]: i think we do not need sampler, when export model

File: tzrec/models/match_model.py [hunk 5]
```diff
@@ -186,20 +205,71 @@ def __init__(
         self._loss_collection = {}
         if self._model_config and hasattr(self._model_config, "in_batch_negative"):
             self._in_batch_negative = self._model_config.in_batch_negative
+        self.sampler_type = kwargs["sampler_type"]
```
Comment (tiankongdeguiji) [inline tzrec/models/match_model.py]: self._sampler_type = kwargs.get("sampler_type", "negative_sampler")

File: tzrec/models/dssm_test.py [hunk 6]
```diff
@@ -115,6 +120,106 @@ def test_dssm(self, graph_type) -> None:
             predictions = dssm(batch)
         self.assertEqual(predictions["similarity"].size(), (2, 3))
 
+    @parameterized.expand(
+        [[TestGraphType.NORMAL], [TestGraphType.FX_TRACE], [TestGraphType.JIT_SCRIPT]]
+    )
+    def test_dssm_hard_negative(self, graph_type) -> None:
+        feature_cfgs = [
```
Comment (tiankongdeguiji) [inline tzrec/models/dssm_test.py:132]: Add an integration test for dssm hard negative

File: docs/source/models/dssm.md [hunk 7]
```diff
@@ -108,7 +108,7 @@ model_config {
 
 ### 负采样配置
```
Comment (tiankongdeguiji) [inline docs/source/models/dssm.md:109]: bump up tzrec version

File: tzrec/datasets/utils.py [hunk 8]
```diff
@@ -170,6 +172,9 @@ def to(self, device: torch.device, non_blocking: bool = False) -> "Batch":
                 k: v.to(device=device, non_blocking=non_blocking)
                 for k, v in self.sample_weights.items()
             },
+            hard_neg_indices=self.hard_neg_indices.to(
```
Comment (tiankongdeguiji) [inline tzrec/datasets/utils.py]: Could hard_neg_indices be optional? so that we do not launch a memcpy kernel for every model. 
By converting `hard_neg_indices` into a dict of tensors (`additional_infos`), we can easily add more tensors without complicating the Batch data structure with additional features.

**Review by tiankongdeguiji — CHANGES_REQUESTED** [review]

File: tzrec/main.py (hunk 3 above)
Comment (eric-gecheng) [reply tzrec/main.py:218]: data_config is a protobuf object, when data_config has no sampler field, the HasField function does not return False, but throw an exception instead.

File: tzrec/models/match_model.py (hunk 5 above)
Comment (eric-gecheng) [reply tzrec/models/match_model.py]: resolved.

File: tzrec/datasets/data_parser.py (hunk 1 above)
Comment (eric-gecheng) [reply tzrec/datasets/data_parser.py:204]: done

File: tzrec/datasets/utils.py (hunk 8 above)
Comment (eric-gecheng) [reply tzrec/datasets/utils.py]: additional_infos dict is added to replace hard_neg_indices.

File: docs/source/models/dssm.md (hunk 7 above)
Comment (eric-gecheng) [reply docs/source/models/dssm.md:109]: 0.7.15

File: tzrec/models/dssm_test.py (hunk 6 above)
Comment (eric-gecheng) [reply tzrec/models/dssm_test.py:132]: done

File: tzrec/datasets/data_parser.py (hunk 2 above)
Comment (eric-gecheng) [reply tzrec/datasets/data_parser.py]: when using JIT_SCRIPT mode, input_data becomes a proxy object, so we have to use hasattr function.

File: tzrec/main.py (hunk 4 above)
Comment (eric-gecheng) [reply tzrec/main.py]: sampler is set to None when export model now.

File: tzrec/datasets/data_parser.py [hunk 9]
```diff
@@ -379,6 +383,15 @@ def to_batch(
         for weight in self._sample_weights:
             sample_weights[weight] = input_data[weight]
 
+        additional_infos = {}
+        hard_neg_indices = torch.tensor([])
```
Comment (tiankongdeguiji) [inline tzrec/datasets/data_parser.py]: may be we do not need the empty tensor?

File: tzrec/datasets/data_parser.py [hunk 10]
```diff
@@ -379,6 +383,15 @@ def to_batch(
...
+        if (
+            hasattr(input_data, "hard_neg_indices")
+            or "hard_neg_indices" in input_data.keys()
+        ):
+            hard_neg_indices = input_data["hard_neg_indices"]
```
Comment (tiankongdeguiji) [inline tzrec/datasets/data_parser.py]: use HARD_NEG_INDICES

File: tzrec/models/dat.py [hunk 11]
```diff
@@ -190,7 +190,12 @@ def predict(self, batch: Batch) -> Dict[str, Tensor]:
         )
 
         ui_sim = (
-            self.sim(user_tower_emb, item_tower_emb) / self._model_config.temperature
+            self.sim(
+                user_tower_emb,
+                item_tower_emb,
+                batch.additional_infos.get("hard_neg_indices", None),
```
Comment (tiankongdeguiji) [inline tzrec/models/dat.py]: use HARD_NEG_INDICES

File: tzrec/models/dssm.py [hunk 12]
```diff
@@ -145,6 +145,11 @@ def predict(self, batch: Batch) -> Dict[str, Tensor]:
         )
 
         ui_sim = (
-            self.sim(user_tower_emb, item_tower_emb) / self._model_config.temperature
+            self.sim(
+                user_tower_emb,
+                item_tower_emb,
+                batch.additional_infos.get("hard_neg_indices", None),
```
Comment (tiankongdeguiji) [inline tzrec/models/dssm.py]: use use HARD_NEG_INDICES

File: tzrec/models/dssm_v2.py [hunk 13]
```diff
@@ -137,6 +137,11 @@ def predict(self, batch: Batch) -> Dict[str, Tensor]:
         item_tower_emb = self.item_tower(item_feat)
 
         ui_sim = (
-            self.sim(user_tower_emb, item_tower_emb) / self._model_config.temperature
+            self.sim(
+                user_tower_emb,
+                item_tower_emb,
+                batch.additional_infos.get("hard_neg_indices", None),
```
Comment (tiankongdeguiji) [inline tzrec/models/dssm_v2.py]: use HARD_NEG_INDICES

File: tzrec/models/mind.py [hunk 14]
```diff
@@ -356,5 +356,10 @@ def predict(self, batch: Batch) -> Dict[str, Tensor]:
         # if self._model_config.similarity == simi_pb2.Similarity.COSINE:
         #     user_emb = F.normalize(user_emb, p=2.0, dim=1)
 
-        ui_sim = self.sim(user_emb, item_emb) / self._model_config.temperature
+        ui_sim = (
+            self.sim(
+                user_emb, item_emb, batch.additional_infos.get("hard_neg_indices", None)
```
Comment (tiankongdeguiji) [inline tzrec/models/mind.py]: use HARD_NEG_INDICES

**Review by tiankongdeguiji — CHANGES_REQUESTED** [review]
